chirp/drivers/baofeng_1900.py

```diff
@@ -0,0 +1,596 @@
+        status.cur = addr + BLOCK_SIZE
+        radio.status_fn(status)
+
+        block = _bf1900_read_block(radio, addr, BLOCK_SIZE)
+        data += block
+
+        LOG.debug("Address: %04x" % addr)
+        LOG.debug(util.hexprint(block))
+
+    _bf1900_exit_programming_mode(radio)
+
+    return memmap.MemoryMapBytes(data)
+
+
+def do_upload(radio):
+    status = chirp_common.Status()
+    status.msg = "Uploading to radio"
+
+    _bf1900_enter_programming_mode(radio)
+
+    status.cur = 0
+    status.max = radio._memsize
+
+    for start_addr, end_addr in radio._ranges:
+        for addr in range(start_addr, end_addr, BLOCK_SIZE):
+            status.cur = addr + BLOCK_SIZE
+            radio.status_fn(status)
+            _bf1900_write_block(radio, addr, BLOCK_SIZE)
+
+    _bf1900_exit_programming_mode(radio)
+
+
+class BaofengBF1904(chirp_common.Alias):
+    VENDOR = 'Baofeng'
+    MODEL = 'BF-1904'
+
+
+@directory.register
+class BF1900Radio(chirp_common.CloneModeRadio):
+    VENDOR = "Baofeng"
+    MODEL = "BF-1901"
+    BAUD_RATE = 9600
+    NEEDS_COMPAT_SERIAL = False
+
+    ALIASES = [BaofengBF1904]
+    SIDEKEYFUNCTION_LIST = ["Off", "Monitor", "Transmit Power", "Alarm"]
+
+    # This code currently requires that ranges start at 0x0000
+    # and are continuous. In the original program 0x0388 and 0x03C8
+    # are only written (all bytes 0xFF), not read.
+    # _ranges = [
+    #       (0x0000, 0x0110),
+    #       (0x02B0, 0x02C0),
+    #       (0x0380, 0x03E0)
+    #       ]
+    # Memory starts looping at 0x1000... But not every 0x1000.
+
+    _ranges = [
+        (0x0000, 0x0110),
+        (0x0380, 0x03E0),
+        (0x02B0, 0x02C0),
+    ]
+    _memsize = 0x03E0
+    _has_fm = True
+    _has_sidekey = False
+    _has_scanmodes = True
+    _has_scramble = False
+
+    def get_features(self):
+        rf = chirp_common.RadioFeatures()
+        rf.has_settings = True
+        rf.valid_modes = ["NFM", "FM"]  # 12.5 KHz, 25 kHz.
+        rf.valid_skips = ["", "S"]
+        rf.valid_tmodes = ["", "Tone", "TSQL", "DTCS", "Cross"]
+        rf.valid_duplexes = ["", "-", "+", "split", "off"]
+        rf.can_odd_split = True
+        rf.has_rx_dtcs = True
+        rf.has_ctone = True
+        rf.has_cross = True
+        rf.valid_cross_modes = [
+            "Tone->Tone",
+            "DTCS->",
+            "->DTCS",
+            "Tone->DTCS",
+            "DTCS->Tone",
+            "->Tone",
+            "DTCS->DTCS"]
+        rf.has_tuning_step = False
+        rf.has_bank = False
+        rf.has_name = False
+        rf.memory_bounds = (1, 16)
+        rf.valid_bands = [(400000000, 520000000)]
+        rf.valid_power_levels = BF1900_POWER_LEVELS
+        rf.valid_tuning_steps = [2.5, 5.0, 6.25, 10.0, 12.5, 15.0, 20.0, 25.0,
+                                 50.0, 100.0]
+
+        return rf
+
+    def process_mmap(self):
+        self._memobj = bitwise.parse(MEM_FORMAT, self._mmap)
```

Comment:
Make this:
```
self._memobj = bitwise.parse(MEM_FORMAT + H777_SETTINGS2, self._mmap)
```
Then override `process_mmap()` in your subclass and use your settings2 variant instead.